spec/abci++/abci++_basic_concepts_002_draft.md

```diff
@@ -1,51 +1,266 @@
+The node providing "chunks" will fetch them from its local Application using
+the `LoadSnapshotChunk` method.
 
-### **Snapshot** connection
+As the new node receives "chunks" it will apply them sequentially to the local
+application with `ApplySnapshotChunk`. When all chunks have been applied, the
+Application's `AppHash` is retrieved via an `Info` query. The `AppHash` is then
+compared to the blockchain's `AppHash` which is verified via
+[light client verification](../light-client/verification/README.md).
 
-* For serving and restoring [state sync snapshots](../abci/apps.md#state-sync).
-* Handles the `ListSnapshots`, `LoadSnapshotChunk`, `OfferSnapshot`, and `ApplySnapshotChunk` calls.
+In summary:
 
-Additionally, there is a `Flush` method that is called on every connection,
-and an `Echo` method that is just for debugging.
+* [**ListSnapshots:**](./abci++_methods_002_draft.md#listsnapshots) Used by nodes to discover available snapshots on peers.
+
+* [**LoadSnapshotChunk:**](./abci++_methods_002_draft.md#loadsnapshotchunk) Used by nodes to retrieve snapshot chunks from peers.
+
+* [**OfferSnapshot:**](./abci++_methods_002_draft.md#offersnapshot) When a node receives a snapshot from a peer, Tendermint uses this method to offer the snapshot to the Application.
+
+* [**ApplySnapshotChunk:**](./abci++_methods_002_draft.md#applysnapshotchunk) Used by Tendermint to hand snapshot chunks to the Application.
+
+### Other methods
+
+Additionally, there is a [**Flush**](./abci++_methods_002_draft.md#flush) method that is called on every connection,
+and an [**Echo**](./abci++_methods_002_draft.md#echo) method that is just for debugging.
 
 >**TODO** Figure out what to do with this.
 
 More details on managing state across connections can be found in the section on
 [ABCI Applications](../abci/apps.md).
 
+## Next-block execution vs. same-block execution
+[&uparrow; Back to Outline](#outline)
+
+With ABCI++ predecessor, ABCI, the only moment when the Application had access to a
+block was when it was decided. This led to a block execution model, called _next-block
+execution_, where some fields hashed in a block header refer to the execution of the
+previous block, namely:
+
+* the merkle root of the Application's state
+* the transaction results
+* the consensus parameter updates
+* the validator updates
+
+With ABCI++, an Application may decide to keep using the next-block execution model;
+however the new methods introduced, `PrepareProposal` and `ProcessProposal` allow
+for a new execution model, called _same-block execution_. An Application implementing
+this execution model, upon receiving a raw proposal via `RequestPrepareProposal`
+and potentially modifying its transaction list,
+fully executes the resulting prepared proposal as though it was the decided block.
+The results of the block execution are used as follows:
+
+* the Application keeps the events generated and provides them if `FinalizeBlock`
```

Comment:
```suggestion
* the Application keeps the events generated and provides them in `FinalizeBlock`
```

I'm not sure what "provides" means here. The application is looking up its saved values from earlier execution, I think?

Reply:
I think this is saying that the application must store the events generated during `PrepareProposal` and finally returns them back to Tendermint during the `FinalizeBlock` call if the block is decided. i.e. there's no way to return events from `PrepareProposal`